src/mame/heathkit/h89.cpp

```diff
@@ -163,23 +297,118 @@ static INPUT_PORTS_START( h89 )
 INPUT_PORTS_END
 
 
+void h89_state::machine_start()
+{
+	save_item(NAME(m_gpp));
+	save_item(NAME(m_rom_enabled));
+	save_item(NAME(m_timer_intr_enabled));
+	save_item(NAME(m_floppy_ram_wp));
+
+
+	u8 *m_floppy_ram_ptr = m_floppy_ram->pointer();
+
+	// Floppy RAM - Read/Only. note: it's not rom, but behaves like it in this
+	// view.
+	m_mem_view[0].install_rom(0x1400, 0x17ff, m_floppy_ram_ptr);
+
+	// Floppy RAM - Read/write
+	m_mem_view[1].install_ram(0x1400, 0x17ff, m_floppy_ram_ptr);
+
+	// update RAM mappings based on RAM size
+	u8 *m_ram_ptr = m_ram->pointer();
+	u32 ram_size = m_ram->size();
+
+	if (ram_size == 0x10000)
+	{
+		// system has a full 64k
+		// All views will have RAM from 8k -> 64k.
+		m_mem_view[0].install_ram(0x2000, 0xffff, m_ram_ptr);
+		m_mem_view[1].install_ram(0x2000, 0xffff, m_ram_ptr);
+		m_mem_view[2].install_ram(0x2000, 0xffff, m_ram_ptr);
+
+		// TODO: install shadow writing to RAM when in ROM mode
+
+		// Only the CP/M - Org 0 view will have it at the lower 8k
+		m_mem_view[2].install_ram(0x0000, 0x1fff, m_ram_ptr + 0xe000);
+	}
+	else
+	{
+		// less than 64k
+
+		// for views with ROM visible, the top of memory is 8k higher than
+		// the memory size, since the base starts at 8k.
+		u32 ram_end = ram_size + 0x1fff;
+
+		m_mem_view[0].install_ram(0x2000, ram_end, m_ram_ptr);
+		m_mem_view[0].unmap_readwrite(ram_end + 1, 0xffff);
+		m_mem_view[1].install_ram(0x2000, ram_end, m_ram_ptr);
+		m_mem_view[1].unmap_readwrite(ram_end + 1, 0xffff);
+
+		// when ROM is not active, memory still starts at 8k
+		m_mem_view[2].install_ram(0x2000, ram_size - 1, m_ram_ptr);
+		// unmap everything above the amount of RAM in the SYSTEM
+		m_mem_view[2].unmap_readwrite(ram_size, 0xffff);
```

Comment:
Once again, install this directly in the address space.  Also, no need to call `unmap_readwrite` when nothing was installed there in the first place.